src/common/charconv.cc

```diff
@@ -0,0 +1,944 @@
+  }
+
+  static int32_t CeilLog2Pow5(const int32_t e) {
+    return RyuPowLogUtils::Log2Pow5(e) + 1;
+  }
+
+  /*
+   * \brief Multiply 32-bit and 64-bit -> 128 bit, then access the higher bits.
+   */
+  static uint32_t MulShift(const uint32_t x, const uint64_t y,
+                           const int32_t shift) noexcept(true) {
+    // For 32-bit * 64-bit: x * y, it can be decomposed into:
+    //
+    //   x * (y_high + y_low) = (x * y_high) + (x * y_low)
+    //
+    // For more general case 64-bit * 64-bit, see https://stackoverflow.com/a/1541458
+    const uint32_t y_low = static_cast<uint32_t>(y);
+    const uint32_t y_high = static_cast<uint32_t>(y >> 32);
+
+    const uint64_t low = static_cast<uint64_t>(x) * y_low;
+    const uint64_t high = static_cast<uint64_t>(x) * y_high;
+
+    const uint64_t sum = (low >> 32) + high;
+    const uint64_t shifted_sum = sum >> (shift - 32);
+
+    return static_cast<uint32_t>(shifted_sum);
+  }
+
+  /*
+   * \brief floor(5^q/2*k) and shift by j
+   */
+  static uint32_t MulPow5InvDivPow2(const uint32_t m, const uint32_t q,
+                                    const int32_t j) noexcept(true) {
+    return MulShift(m, kFloatPow5InvSplit[q], j);
+  }
+
+  /*
+   * \brief floor(2^k/5^q) + 1 and shift by j
+   */
+  static uint32_t MulPow5divPow2(const uint32_t m, const uint32_t i,
+                                 const int32_t j) noexcept(true) {
+    // clang-tidy makes false assumption that can lead to i >= 47, which is impossible.
+    // Can be verified by enumerating all float32 values.
+    return MulShift(m, kFloatPow5Split[i], j);  // NOLINT
+  }
+
+  static uint32_t FloorLog2(const uint64_t value) {
+    return 63 - __builtin_clzll(value);
+  }
+
+  /*
+   * \brief floor(e * log_10(2)).
+   */
+  static uint32_t Log10Pow2(const int32_t e) noexcept(true) {
+    // The first value this approximation fails for is 2^1651 which is just
+    // greater than 10^297.
+    assert(e >= 0);
+    assert(e <= 1 << 15);
+    return static_cast<uint32_t>((static_cast<uint64_t>(e) * 169464822037455ull) >> 49);
+  }
+
+  // Returns floor(e * log_10(5)).
+  static uint32_t Log10Pow5(const int32_t expoent) noexcept(true) {
+    // The first value this approximation fails for is 5^2621 which is just
+    // greater than 10^1832.
+    assert(expoent >= 0);
+    assert(expoent <= 1 << 15);
+    return static_cast<uint32_t>(
+        ((static_cast<uint64_t>(expoent)) * 196742565691928ull) >> 48);
+  }
+};
+
+constexpr uint64_t RyuPowLogUtils::kFloatPow5InvSplit[55];
+constexpr uint64_t RyuPowLogUtils::kFloatPow5Split[47];
+
+class PowerBaseComputer {
+ private:
+  static uint8_t
+  ToDecimalBase(bool const accept_bounds, uint32_t const mantissa_low_shift,
+                MantissaInteval const base2, MantissaInteval *base10,
+                bool *mantissa_low_is_trailing_zeros,
+                bool *mantissa_out_is_trailing_zeros) noexcept(true) {
+    uint8_t last_removed_digit = 0;
```

Comment:
Just a note that if there is a problem with code in functions like this it will be extremely difficult and time-consuming for current maintainers to fix. This is because the code is highly specialised and requires domain knowledge.

Reply:
@RAMitchell I agree. When I was reviewing this PR, I just treated `charcov.cc` as a black-box.

It all comes down to the question of whether we should implement a feature ourselves or use a third-party library. In the original RFC discussion for JSON, we decided against including third-party libraries for JSON, e.g. see comment https://github.com/dmlc/xgboost/issues/3916#issuecomment-443273481. The reason was that we wanted to build XGBoost for multiple platforms and languages, and existence of 3rd-party dependencies like RapidJSON would make that difficult.

If we really do not want to include `charconv` in the XGBoost codebase, here are some alternatives:
* Statically link the original Ryu library. This is rather difficult, since it uses Bazel for building.
* Statically link the [fmtlib library](https://github.com/fmtlib/fmt). This library is header-only, so can be used with both CMake and Makefile. The library is slated to become part of C++20, and I've had great experience using this library in Treelite.

Reply:
1 is a no go.  It's more like a reference implementation, I don't think it's intended for these kind of usage.  How does fmtlib parse the value back?  I thought it's just printing.  If other maintainer really want to use 3 party options I'm happy to let him choose the next steps.

Reply:
> How does fmtlib parse the value back?

It does not. We can keep the JSON parser implementation (+ parsing) and just use fmtlib for printing. My understanding is that the specialized implementation is mostly the printing of floats.

My opinion is that a large third party libs like RapidJSON is a no go. I can accept something like fmtlib. But I’d rather get charconv merged and augment it with more documentation.

Let’s wait to hear from @RAMitchell.

Reply:
The C++ ecosystem is immensely decentralized with many competing build systems: Make, CMake, Autotools, Bazel, Meson, build2 and others. This makes a great pain to integrate with third party libs, especially because XGBoost already supports two build systems (Autotools for R, CMake for all else) and DMLC-Core supports three (+GNU Make). So if a third-party lib works with one build system but not with another, we cannot use it.

Reply:
You need to have a correctly rounding algorithm for parsing number, which is also included in this PR (part of ryu , but I shrink the double table to float table). So it's best to have a roundtrip implementation.  Besides, since the functions here are defined by C++ 17,  we can eventually use that.

Reply:
Ah so Ryu is really a round trip algorithm?

Reply:
Recently yes.

Reply:
I'm okay to continue with this is the other options we have discussed are not great. I only mention this to be explicit about the risks we are taking, and that we should try to avoid 'black box' situations where possible otherwise identify them and try to mitigate the problems.

Reply:
Got it. I should make time to actually read the Ryu paper then. Let’s see if we can add more fine grained unit tests for subcomponents.

Reply:
I think it's not too bad.  I understand the concern, that's why I went ahead and just enumerate all floating point numbers to ensure it's correctness.  Maybe it will have bug in arm or some other places that we don't have test but other than that it's 100 coverage.

Reply:
@trivialfis Can you add the code for the brute force test? Is it written in pytest or Google Test? The CI won't run it but we can run it manually from time to time.

Reply:
@hcho3 Change this line:

https://github.com/dmlc/xgboost/blob/38ee51478740cb1aae1cfa0791168d384c388082/tests/cpp/common/test_json.cc#L548

to `i+=1`.  Right now on master branch I use sampling to reduce the test time.

Reply:
Thanks. Let me submit a small PR so that we can optionally enable the brute force test.